File: backend/models/Product.js
const mongoose = require('mongoose');

const productSchema = new mongoose.Schema({
    name: {
        type: String,
        required: [true, 'Product name is required'],
        trim: true,
        maxlength: [100, 'Product name cannot exceed 100 characters']
    },
    description: {
        type: String,
        required: [true, 'Product description is required'],
        maxlength: [1000, 'Description cannot exceed 1000 characters']
    },
    price: {
        type: Number,
        required: [true, 'Product price is required'],
        min: [0, 'Price cannot be negative']
    },
    category: {
        type: String,
        required: [true, 'Product category is required'],
        enum: ['electronics', 'fashion', 'home', 'books']
    },
    images: [{
        type: String,
        validate: {
            validator: function(v) {
                return /^https?:\/\//.test(v);
            },
            message: 'Image URL must be a valid URL'
        }
    }],
    sellerId: {
        type: Number,
        required: [true, 'Seller ID is required']
    },
    sellerName: {
        type: String,
        required: [true, 'Seller name is required']
    },
    stock: {
        type: Number,
        required: [true, 'Stock quantity is required'],
        min: [0, 'Stock cannot be negative'],
        default: 0
    },
    status: {
        type: String,
        enum: ['active', 'inactive', 'pending'],
        default: 'active'
    },
    views: {
        type: Number,
        default: 0
    }
}, {
    timestamps: true
});

// Index for better search performance
productSchema.index({ name: 'text', description: 'text' });
productSchema.index({ sellerId: 1 });
productSchema.index({ category: 1 });
productSchema.index({ price: 1 });
productSchema.index({ createdAt: -1 });

module.exports = mongoose.model('Product', productSchema);

File: backend/models/product.py
from pydantic import BaseModel, validator
from typing import Optional, List
from datetime import datetime
import uuid

class ProductCreate(BaseModel):
    name: str
    description: str
    price: float
    category: str
    images: List[str] = []
    video_url: Optional[str] = None
    stock: int = 1
    tags: List[str] = []
    weight: Optional[float] = None  # in kg for shipping calculations
    dimensions: Optional[dict] = None  # {"length": 0, "width": 0, "height": 0} in cm

    @validator('price')
    def validate_price(cls, v):
        if v <= 0:
            raise ValueError('Price must be greater than 0')
        return v

    @validator('stock')
    def validate_stock(cls, v):
        if v < 0:
            raise ValueError('Stock cannot be negative')
        return v

class ProductUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    category: Optional[str] = None
    images: Optional[List[str]] = None
    video_url: Optional[str] = None
    stock: Optional[int] = None
    tags: Optional[List[str]] = None
    weight: Optional[float] = None
    dimensions: Optional[dict] = None

class ProductResponse(BaseModel):
    id: str
    name: str
    description: str
    price: float
    category: str
    images: List[str]
    video_url: Optional[str] = None
    stock: int
    tags: List[str]
    weight: Optional[float] = None
    dimensions: Optional[dict] = None
    seller_id: str
    seller_name: str
    views: int = 0
    is_active: bool = True
    created_at: datetime
    updated_at: datetime

class ProductInDB(BaseModel):
    id: str
    name: str
    description: str
    price: float
    category: str
    images: List[str]
    video_url: Optional[str] = None
    stock: int
    tags: List[str]
    weight: Optional[float] = None
    dimensions: Optional[dict] = None
    seller_id: str
    seller_name: str
    views: int = 0
    is_active: bool = True
    created_at: datetime
    updated_at: datetime